File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque
def solution(bridge_length, weight, truck_weights):
    
    time = 0
    bridge = deque([0] * bridge_length)
    truck_weights = deque(truck_weights)
    
    currentWeight = 0
    while len(truck_weights)!=0:
        time+=1

        currentWeight -= bridge.popleft()

        if currentWeight + truck_weights[0] <= weight:
            currentWeight+= truck_weights[0]
            bridge.append(truck_weights.popleft())

        else: 
            bridge.append(0)
            
    time += bridge_length
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
  const waiting = [...truck_weights];
  const bridgeQueue = [];
  let time = 0;
  let onBridge = 0;

  while (waiting.length > 0 || onBridge > 0) {
    time += 1;
    if (bridgeQueue.length && bridgeQueue[0].exit === time) {
      onBridge -= bridgeQueue[0].w;
      bridgeQueue.shift();
    }
    if (waiting.length && onBridge + waiting[0] <= weight && bridgeQueue.length < bridge_length) {
      const w = waiting.shift();
      onBridge += w;
      bridgeQueue.push({ w, exit: time + bridge_length });
    }
  }

  return time;
}